packages/babel-plugin-proposal-decorators/test/fixtures/legacy-class-inner-refs/exec.mjs

```diff
@@ -0,0 +1,42 @@
+let Child
+
+function dec(cls) {
+  Child = class extends cls {}
+  return Child
+}
+
+function dec2(cls) {
+  return class extends cls {}
+}
+
+@dec
+@dec2
+class Parent {
+  constructor() {
+    this.prop = Parent
+  }
+
+  static static() {
+    return Parent
+  }
+
+  method() {
+    return Parent
```

Comment:
Can you add test cases on `property`? i.e.
```js
@dec
@dec2
class Parent {
  property = Parent,
}
```
What will `p.property` be? Same for private properties.

Reply:
>Solved by inserting in all methods (including static, #private ones and constructor) let ClassName = _someGeneratedVariable, where _someGeneratedVariable hold "final" class, after all decorators applied.
>References in class properties still isn't supported.

Property isn't working currently (will hold undecorated class). For full solution we should replace all class references in read positions of class body to outer reference, but I ain't good enough at babel for such tricks.

Reply:
Hi! Catch issue with property yesterday, similar to @JLHwung  case, but with static property)

example source:
```
function dec(cls) {
  return cls
}

@dec
class Decorated {
  static property = Decorated
}
```

and output:
```
"use strict";

var _class, _class2, _temp;

function dec(cls) {
  return cls;
}

let Decorated = dec(_class = (_temp = _class2 = class Decorated {}, _class2.property = Decorated, _temp)) || _class;
```

This code produce runtime error `ReferenceError: Cannot access 'Decorated' before initialization`

Reply:
@Amareis about class properties, fix references require changes in `@babel/plugin-proposal-class-properties` plugin?

Reply:
> @Amareis about class properties, fix references require changes in `@babel/plugin-proposal-class-properties` plugin?

Don't think so. If we get full solution with replacing of original class reference in class body, this will be working without any additional efforts.